sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/CbsTokenProvider.cs

```diff
@@ -62,14 +83,94 @@ internal sealed class CbsTokenProvider : ICbsTokenProvider
         /// <param name="appliesTo">The resource to which the token should apply.</param>
         /// <param name="requiredClaims">The set of claims that are required for authorization.</param>
+        ///
         /// <returns>The token to use for authorization.</returns>
         ///
         public async Task<CbsToken> GetTokenAsync(Uri namespaceAddress,
                                                   string appliesTo,
-                                                  string[] requiredClaims)
+                                                  string[] requiredClaims) =>
+            Credential switch
+            {
+                _ when Credential.IsSharedAccessCredential => await AcquireSharedAccessTokenAsync().ConfigureAwait(false),
+                _ => await AcquireJwtTokenAsync().ConfigureAwait(false)
+            };
+
+        /// <summary>
+        ///   Performs the task needed to clean up resources used by the <see cref="CbsTokenProvider" />.
+        /// </summary>
+        ///
+        public void Dispose() => TokenAcquireSemaphore?.Dispose();
+
+        /// <summary>
+        ///   Acquires a token for a JWT-type credential.
+        /// </summary>
+        ///
+        /// <returns>The <see cref="CbsToken"/> to use for authorization.</returns>
+        ///
+        private async Task<CbsToken> AcquireJwtTokenAsync()
+        {
+            var token = Volatile.Read(ref _cachedJwtToken);
+
+            // If the token is expiring, attempt to acquire the semaphore needed to
+            // refresh it.
+
+            if (IsExpiring(token))
+            {
+                try
+                {
+                    // Attempt to acquire the semaphore synchronously, in case it is not currently
+                    // held.
+
+                    if (!TokenAcquireSemaphore.Wait(0, CancellationToken.None))
+                    {
+                        await TokenAcquireSemaphore.WaitAsync(CancellationToken).ConfigureAwait(false);
+                    }
+
+                    // Because the token may have been refreshed while waiting for the semaphore,
+                    // check expiration again to avoid making an necessary request.  Because the token
+                    // is only updated when the semaphore is held, there's no need to perform a volatile
+                    // read for this refresh.
+
+                    token = _cachedJwtToken;
+
+                    if (IsExpiring(token))
+                    {
+                        var accessToken = await Credential.GetTokenUsingDefaultScopeAsync(CancellationToken).ConfigureAwait(false);
+                        token = new CbsToken(accessToken.Token, TokenType, accessToken.ExpiresOn.UtcDateTime);
+
+                        _cachedJwtToken = token;
+                    }
+                }
+                finally
+                {
+                    TokenAcquireSemaphore.Release();
```

Comment:
Do we need to guard this release, in case the semaphore throws?

Reply:
I don't believe so.    I'll double-check the behavior if we don't acquire, but I believe that it is safe.   If not, I'll add a guard.

Reply:
We usually add a guard in this situation, because Release throws if the lock was not acquired.

Reply:
Just double-checked.  You're right... we need the guard.